flink-optimizer/src/main/java/org/apache/flink/optimizer/operators/GroupReduceWithCombineProperties.java

```diff
@@ -87,19 +92,39 @@ public DriverStrategy getStrategy() {
 		return DriverStrategy.SORTED_GROUP_REDUCE;
 	}
 
-	@Override
 	public SingleInputPlanNode instantiate(Channel in, SingleInputNode node) {
 		if (in.getShipStrategy() == ShipStrategyType.FORWARD) {
 			// adjust a sort (changes grouping, so it must be for this driver to combining sort
-			if (in.getLocalStrategy() == LocalStrategy.SORT) {
-				if (!in.getLocalStrategyKeys().isValidUnorderedPrefix(this.keys)) {
-					throw new RuntimeException("Bug: Inconsistent sort for group strategy.");
+			if(in.getSource().getOptimizerNode() instanceof PartitionNode) {
+				// Inject a combiner before the partition node
+				Channel toCombiner = new Channel(in.getSource());
+				toCombiner.setShipStrategy(ShipStrategyType.FORWARD, DataExchangeMode.PIPELINED);
+				GroupReduceNode combinerNode = ((GroupReduceNode) node).getCombinerUtilityNode();
+				combinerNode.setParallelism(in.getSource().getParallelism());
+				if(toCombiner.getSource().getInputs().iterator().hasNext()) {
+					Channel source = toCombiner.getSource().getInputs().iterator().next();
+					// A combiner plan node is created with the map as the input
+					SingleInputPlanNode combiner = new SingleInputPlanNode(combinerNode, "Combine("+node.getOperator()
```

Comment:
Thanks for all the comments. Checking one by one. I have few questions. Let me get back to you after fixing the other comments. Sorry about the formatting things. I will take care from next time so that they are not repeated.